packages/common/src/hooks/usePremiumContentAccess.ts

```diff
@@ -0,0 +1,60 @@
+import { PremiumConditions, Track } from 'models'
+import { useMemo } from 'react'
+
+import { useSelector } from 'react-redux'
+import { cacheUsersSelectors } from 'store/cache'
+import { premiumContentSelectors } from 'store/premium-content'
+import { CommonState } from 'store/reducers'
+import { Nullable } from 'utils'
+
+const { getUsers } = cacheUsersSelectors
+const { getPremiumTrackSignatureMap } = premiumContentSelectors
+
+export const usePremiumContentAccess = (track: Nullable<Partial<Track>>) => {
+  if (!track) {
+    return { isUserAccessTBD: false, doesUserHaveAccess: true }
+  }
+
+  const premiumTrackSignatureMap = useSelector(getPremiumTrackSignatureMap)
+
+  const { isUserAccessTBD, doesUserHaveAccess } = useMemo(() => {
+    const trackId = track.track_id
+    const isPremium = track.is_premium
+    const hasPremiumContentSignature =
+      !!track.premium_content_signature ||
+      !!(trackId && premiumTrackSignatureMap[trackId])
+    const isCollectibleGated = !!track.premium_conditions?.nft_collection
+    const isSignatureToBeFetched =
+      isCollectibleGated &&
+      !!trackId &&
+      premiumTrackSignatureMap[trackId] === undefined
+
+    return {
+      isUserAccessTBD: !hasPremiumContentSignature && isSignatureToBeFetched,
+      doesUserHaveAccess: !isPremium || hasPremiumContentSignature
+    }
+  }, [track, premiumTrackSignatureMap])
+
+  return { isUserAccessTBD, doesUserHaveAccess }
+}
+
+export const useSpecialAccessEntity = (premiumConditions: Nullable<PremiumConditions>) => {
+  const { follow_user_id: followUserId, tip_user_id: tipUserId } =
+    premiumConditions ?? {}
+
+  const users = useSelector((state: CommonState) =>
+    getUsers(state, {
+      ids: [followUserId, tipUserId].filter((id): id is number => !!id)
```

Comment:
Not quite following this, my understanding is `id is number` proves that `id` is of type `number`? Can't just use `removeNullable` or something?
is it bc `followerUserId` comes from the backend and we don't know the type yet?

Reply:
it's a neat little trick that says that the ids that satisfy the filter condition should be considered as number. otherwise TS may complain and still think that the id, even after filter, could be undefined/null

Reply:
does our `removeNullable` type helper not work?